### PR TITLE
Enable Fast Double Parser in Jackson

### PR DESCRIPTION
**Pull request by mgodwan**
### Description
Enable Jackson Features `USE_FAST_DOUBLE_PARSER` and `USE_FAST_DOUBLE_WRITER`

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/7822

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Verified all the content types with `_doc` API:

1. Smile
2. YAML
3. CBOR
4. JSON

JSON Document:
```
{ "account_number": 1,"address": "Some Avenue", "lat": 71.38  }
```

YAML Document:

```
---
account_number: 1
address: Some Avenue
lat: 71.38
```

### CODE REVIEW

**Comment by dblock** [comment]
Let's add to the CHANGELOG something like "Improved performance of parsing ..."?

**Comment by reta** [comment]
@dblock @mgodwan I was looking into this feature and there are few "dragons" here:

 - https://github.com/FasterXML/jackson-core/issues/970 (takeaway, with Java 17 there is negligible speed up of 1-2%)
 - https://github.com/FasterXML/jackson-core/issues/577 (takeaway, the results of the conversion may differ from the JDK)

I think this is a good feature to offer but may be we should be careful with using it as default?

**Comment by mgodwan** [comment]
@reta I ran a few benchmarks with nyc taxis where I saw the following differences in throughput without and with this change:


Code Path | Indexing Throughput Avg | P50 Indexing Latency |
|-|-|-|
Baseline | 78162.21,docs/s | 876.34 ms |
Candidate (Fast Double Parsing Enabled) | 81743.31,docs/s (1.04x)  | 837.26 ms (0.956x) |

These runs use JDK17 and show good improvements.

File: libs/x-content/src/main/java/org/opensearch/common/xcontent/json/JsonXContent.java [hunk 1]
```diff
@@ -78,6 +78,8 @@ public static XContentBuilder contentBuilder() throws IOException {
         jsonFactory.configure(JsonGenerator.Feature.AUTO_CLOSE_JSON_CONTENT, false);
         jsonFactory.configure(JsonParser.Feature.STRICT_DUPLICATE_DETECTION, true);
         jsonFactory.setStreamReadConstraints(StreamReadConstraints.builder().maxStringLength(DEFAULT_MAX_STRING_LEN).build());
+        jsonFactory.configure(JsonParser.Feature.USE_FAST_DOUBLE_PARSER, true);
```
[the remark this answers is not in the document]
Comment (mgodwan) [reply libs/x-content/src/main/java/org/opensearch/common/xcontent/json/JsonXContent.java]: Updated.

**Comment by reta** [comment]
> @reta I ran a few benchmarks with nyc taxis where I saw the following differences in throughput without and with this change:

Thanks @mgodwan , this is obviously macro benchmarks (vs microbenchmarks shared by project maintainers) but the improvements for this specific workload are in more or less intact for JDK-17.  

It would be great to confirm there is no differences in values produced by JDK (current) and `FastDoubleParsing` at least for this workload (the other concern raised by project maintainers).

**Comment by mgodwan** [comment]
> Thanks @mgodwan , this is obviously macro benchmarks (vs microbenchmarks shared by project maintainers) but the improvements for this specific workload are in more or less intact for JDK-17.

Agreed that this would be workload dependent. I chose nyc taxis as it includes certain geo-point fields which are parsed as floating point numbers so they give a good indication of the kind of improvement we can expect. I had executed some micro-benchmarks on JDK17 before opening the issue for conversion from String to Double data types which yielded gains as well.


> It would be great to confirm there is no differences in values produced by JDK (current) and FastDoubleParsing at least for this workload (the other concern raised by project maintainers). 

I'll see how we can verify this for the workload. There are 2 aspects to this confirmation:

1. For parsing, Jackson inlines [github.com/wrandelshofer/FastDoubleParser](https://github.com/wrandelshofer/FastDoubleParser) which is a Java port of github.com/fastfloat/fast_float and the algorithm used provides exact parsing according to the IEEE standard (https://arxiv.org/abs/2101.11408) I don't think there is any concern raised for this aspect on the other issue you had cited on the precision.

2. For serializing, Jackson uses Schubfach's algorithm which seems to have been adopted in the latest versions of [openjdk at least](https://github.com/openjdk/jdk/commit/72bcf2aa03d53b0f68eb07a902575b4e8628d859) as well with the implementation by the same author. Interestingly, Schubfach's provide correct serialization for a few corner cases which account for the difference which may be seen (see: https://bugs.openjdk.org/browse/JDK-4511638) and I could verify this by running against JDK17 and JDK20 for a few cases mentioned in the openjdk bug. We may benefit by allowing the use of Jackson's feature for OpenSearch 2.x which still uses JDK17 and does not have this implementation within JDK.

**Comment by mgodwan** [comment]
1. I've removed the USE_FAST_DOUBLE_WRITER from this PR since JDK20 already takes care of it.
2. For 2.x backport, I can modify the PR to include the feature `USE_FAST_DOUBLE_WRITER` so that we can handle the corner cases in earlier JDK implementations as well.

Please take a look at the PR and let me know your thoughts. 

cc: @reta @dblock @backslasht

**Review by backslasht — APPROVED** [review]
Thanks @mgodwan! Looks good to me.

**Comment by backslasht** [comment]
> Thanks @mgodwan , I think making it default for 3.x is reasonable, for 2.x would be good to have a way to let users opt-out (system property for example).

@reta - Are you suggesting to provide an opt-out mechanism as there is a change in behavior? 

> On the related note, we use Jackson for CBOR/Smile/Yaml (CborXContent/SmileXContent/YamlXContent), should we also switch those to USE_FAST_DOUBLE_WRITER? I think it would make sense.

+1

**Comment by reta** [comment]
> @reta - Are you suggesting to provide an opt-out mechanism as there is a change in behavior?

Correct, we bundle JDK-17 with 2.x but users could (and do) use any JDK-11+ distribution, in case any issues are encountered, we would have a mechanism to hand off parsing back to JDK

**Comment by mgodwan** [comment]
> On the related note, we use Jackson for CBOR/Smile/Yaml (CborXContent/SmileXContent/YamlXContent), should we also switch those to USE_FAST_DOUBLE_WRITER? I think it would make sense.

I have updated the parser configuration for CborXContent, SmileXContent, and YamlXContent as well.

**Comment by reta** [comment]
@mgodwan conflicts :(

**Comment by mgodwan** [comment]
Resolved the conflicts.

**Comment by backslasht** [comment]
@mgodwan - Can you please add testing notes for the non JSON formats?

**Review by reta — APPROVED** [review]
@dblock are we good with making it default for `main` and `2.x` but have an opt-out option for 2.x users (system property)?

**Review by backslasht — APPROVED** [review]

**Comment by dblock** [comment]
> @dblock are we good with making it default for `main` and `2.x` but have an opt-out option for 2.x users (system property)?

For 3.0 (main) looks good.

For 2.x, does it actually change behavior, even on edge cases? If so I would say it's a no-go because it would break semver. It should be optional.

**Comment by mgodwan** [comment]
Thanks @dblock @reta 

Could you help with merging this to `main`? 
I can take up the required changes for backport once workflow is triggered.

**Comment by reta** [comment]
> Could you help with merging this to `main`?

@dblock sign off?

**Comment by dblock** [comment]
I merged it, however I am still unclear what the edge cases are. I would at least add a test that demonstrates the handling of the edge case (a test that would fail if we remove this code).

**Comment by reta** [comment]
@dblock there is a difficult one (at least it seems to me):
 - we could look for conversion edge case (still have to find some), where JDK conversion != FastDoubleConversion, that could actually change from JDK to JDK (from the discussions above)
- we could look for conversion time edge case (more like JMH-oriented test cases) where we assert that FastDoubleConversion should not take longer than X 

Which ones you are thinking about? Or something very different?

**Comment by mgodwan** [comment]
> we could look for conversion edge cases (still have to find some), where JDK conversion != FastDoubleConversion, that could actually change from JDK to JDK (from the discussions above)

For serialization, some cases which yield different output on jdk-17 and with jackson flag enabled are listed [here](https://bugs.openjdk.org/browse/JDK-4511638).

e.g.
```
JDK 17 behaviour: 8.41E21 --> 8.409999999999999E21
JDK 20 behaviour: 8.41E21 --> 8.41E21
Jackson flag behaviour: 8.41E21 --> 8.41E21
```

I could not find any cases for JDK-20 since jdk-20 and jackson seem to be using the same algorithm for converting java double to string.

**Comment by mgodwan** [comment]
> I merged it, however I am still unclear what the edge cases are. I would at least add a test that demonstrates the handling of the edge case (a test that would fail if we remove this code).

* In this PR, we have only enabled jackson flag for converting from string to double[uses `USE_FAST_DOUBLE_PARSER`].
* We have not enabled jackson implementation for double to string conversion [uses `USE_FAST_DOUBLE_WRITER`]. 
* The string to double procedure does not have any edge cases we are aware of (the algorithm follows IEEE standard) while there are edge cases in highlighted in double to string are covered in the above comment.

**Comment by dblock** [comment]
@mgodwan So you're saying that the two outcomes of this change are 1) improved performance, and 2) fixing a bug such as specifying 8.41E21 would produce 8.409999999999999E21? If so YOLO.

**Pull request merged**